### ISSUE TITLE
Support pushing to insecure http-only registry

### ISSUE THREAD

**Issue by sipsma** [dagger/dagger#4007]
User request to be able to push to http registry: https://discord.com/channels/707636530424053791/1046817210695417906/1046817210695417906

I can confirm that I get the same error but only when trying to push to a non-https registry that is not on localhost. There must be some logic buried somewhere in buildkit or its image push dependencies to not enfore https on localhost?

Either way, buildkit actually has a flag to support this during image export but for some reason I can't get it to work; when I run a locally modified test case to push to an ip on my local network then I still get the same `server gave HTTP response to HTTPS client` error... Code here: https://github.com/sipsma/dagger/commit/4314f6559dadfe3bd441b44eb9bd4460501fc7eb

Not sure if I'm doing something wrong or if this isn't working upstream (I did notice that it doesn't appear to be tested anywhere upstream). cc @vito in case you have had to deal with this setting in the past and know any more.

**Comment by vito** [comment]
> There must be some logic buried somewhere in buildkit or its image push dependencies to not enfore https on localhost?

Yep, I ran into similar logic for resolving, so it wouldn't surprise me if pushing followed the same rules:

* https://github.com/moby/buildkit/blob/4df3e8f39c851a23c7d9c43a803b36cf9cab76e3/util/resolver/pool.go#L100
* https://github.com/moby/buildkit/blob/4df3e8f39c851a23c7d9c43a803b36cf9cab76e3/util/resolver/resolver.go#L33
* https://github.com/moby/buildkit/blob/4df3e8f39c851a23c7d9c43a803b36cf9cab76e3/util/resolver/resolver.go#L167

I never found a way to convince Buildkit to use plain HTTP for non-localhost addresses. :confused: This is what sent me down the path of having Bass generate TLS certs for its services (since I wanted to run a registry as a service).

I think Buildkit's notion of "insecure" just means it doesn't verify TLS certs, not "don't use TLS."

**Comment by sipsma** [comment]
Thanks @vito, the localhost part makes sense, I'm still confused though because the code *seems* to very heavily imply that it's trying to allow plain http here: https://github.com/sipsma/buildkit/blob/0ebdec875d2377ea493c082be46bc28b7411f5e6/util/push/push.go#L78-L78

I asked in the buildkit slack, will update
